scripts/safe_template_linter.py

```diff
@@ -654,6 +667,44 @@ def _check_deprecated_display_name(self, expression, results):
                 Rules.mako_deprecated_display_name, expression
             ))
 
+    def _check_html_and_text(self, expression, results):
+        """
+        Checks rules related to proper use of HTML() and Text().
+
+        Rule 1: If HTML() is called, the expression must begin with Text().
+        Rule 2: Both HTML() and Text() must be closed before any call to
+            format().
+        Rule 3: Using Text() without HTML() is unnecessary.
+
+        Arguments:
+            expression: A dict containing the start_index, end_index, and
+                expression (text) of the expression.
+            results: A list of results into which violations will be added.
+
+        """
+        # strip '${' and '}' from ends
+        expression_inner = expression['expression'][2:-1]
+        if 'HTML(' in expression_inner:
+            if expression_inner.lstrip().startswith('Text(') is False:
+                results.violations.append(ExpressionRuleViolation(
+                    Rules.mako_html_requires_text, expression
+                ))
+        else:
+            if 'Text(' in expression_inner:
+                results.violations.append(ExpressionRuleViolation(
+                    Rules.mako_text_redundant, expression
+                ))
+
+        for match in re.finditer("(HTML\(|Text\()", expression_inner):
+            close_paren_index = self._find_closing_char_index(None, "(", ")", expression_inner, match.end(), 0)
+            if 0 <= close_paren_index:
+                # the argument sent to HTML() or Text()
+                argument = expression_inner[match.end():close_paren_index]
```

Comment:
@efischer19: The call _find_closing_char_index is a bit smarter than regex because it skips matching parens that appear inside.  For example, if you had something like the following, it would find the correct closing paren:
```
Text(_("some text"))
```
Regex can't be as smart I don't think.